### ISSUE TITLE
what is sampler?

### ISSUE THREAD

**Issue by songhune** [NESAP-STL/nesap-stl#8]
dataset--__init__.py / line 25~
`# Distributed samplers sampler: 
    train_sampler, valid_sampler = None, None
    if distributed and use_dist_sampler_train:
        train_sampler = DistributedSampler(train_dataset)
    if distributed and use_dist_sampler_valid and valid_dataset is not None:
        valid_sampler = DistributedSampler(valid_dataset)
`

**Comment by songhune** [comment]
torch.utils.data.Dataset에 정의된 데이터 class를 알 필요가 있습니다.
1. Map-style dataset
index가 존재하여 data[index]로 데이터를 참조할 수 있음
__getitem__과 __len__ 선언 필요
2. 여기에서, torch.utils.data.Sampler를 사용하는데, 이는 
SequentialSampler : 항상 같은 순서
RandomSampler : 랜덤, replacemetn 여부 선택 가능, 개수 선택 가능
SubsetRandomSampler : 랜덤 리스트, 위와 두 조건 불가능
WeigthRandomSampler : 가중치에 따른 확률
BatchSampler : batch단위로 sampling 가능
DistributedSampler : 분산처리 (torch.nn.parallel.DistributedDataParallel과 함께 사용)
이 중에서 DistributedSampler가 여기서는 사용됩니다.

**Comment by ks1101** [comment]
별로 쓸모 없을 지도 모르지만 한게 아깝기 때문에...

sea 데이터 기준으로 했습니다. climate도 같을지도..?
1.get_data_loader	//데이터를 train_data_loader, valid_data_loader로 분리

get_data_loader	(datasets/_init_.py)
	get_datasets	(datasets/climate.py)
		1.데이터 로드
		2.데이터 크기 조정
		data size   -> (14389,70,125) = (T,H,W) ==(time, height, width)??
                    -> (14389,68,120) //124로 안 하고 120으로 자른 이유가 있나요? patch_size==4라 124도 될 것 같은데
                    -> (1,14389,1,68,120) // reshape_patch 돌리려고, dummy batch dim,channel dim 추가
        reshape_patch   (utils/preprocess.py)
            patch_size=4
            data    -> (1,14389, 1, 68, 120)
                    -> (1, 14389, 16, 17, 30)   //channel_dim *= 4*4, W/=4, T/=4
                                                //channel_dim이 확실한 건 아니고 추측하는 겁니다.
                                                //channel을 여기서 적용하는 게 맞는지, 왜 적용하는 건지는 잘 모르겠습니다.
            return data 크기 조정한 결과
        
        get_datasets cont.	(datasets/climate.py)
        n_train_seq = 10255 = 10240+16-1
        train_data  -> data[0:10255]        //10256
        valid_data  -> data[10256:14388]    //4133
        train_set = train_data + seq_len(==16)
        valid_set = valid_data + seq_len(==16)
        
        return train_set, valid_set, {} //{}는 loader_args에 해당
    
    get_data_loader	cont. (datasets/_init_.py)
    train_sampler, valid_sampler    //sampler distribution_sampler 사용
                                    //sampler가 학습할 때 sample을 분산처리? 해서 제공하는 역할인 것 같은데 맞나요?
    train_loader = DataLoader(train_dataset, batch_size, sampler, shuffle, **loader_args)
    valid_loader = DataLoader(valid_dataset, batch_size, sampler, **loader_args)
    //train_datasets -> (10256,16,17,30), seq_len = 16
    //batch_size = 4
    //sampler = train_sampler = distribution_sampler?
    //shuffle =T,F?
    //loader_args = {}  loader_args는 어떤 역할?
    
    return train_loader, valid_loader

의문점
1. reshape_size 전에 data를 (14389, 68, 124)가 아니라 (14389,68,120)으로 자른 이유가 있나요?
2. reshape_size 결과인 data -> (14389, 16, 17, 30)에서 16이 channel_dim이 맞는지? channel은 정확히 어떤 역할?
3. loader_args는 DataLoader 찾아보니까 batch_sample, num_workers, collate_fn등을 말하는 것 같은데 맞나요?

**Comment by ks1101** [comment]
2. get_trainer //   trainers/auto_regressive.py, basic.py, base.py 다니면서 trainer 정보 취득
    trainer/base.py에 evaluate, train_epoch등 미구현 되있다고 말씀하셨는데, 
    basic.py나 auto_regressive에 구현이 되어 있어서 신경 안 써도 될 것 같기도 합니다.

    return AutoRegressiveTrainer    //class임, basic,base 정보 포함

3. trainer.build
    base.py에는 미구현이고 basic.py에 있는데, 
    model = predrnn_pp, optimizer = adam, loss = MSE등의 정보를 가져오는 것 같습니다.
    
    근데 metrics는 뭔지 모르겠네요. predrnn_climate.yaml에 없는 내용이라
    utils/metrics.py가 있기는 한데 잘 모르겠네요.
    애초에 metrics 보고 행렬이라고 생각하고 있었을 정도의 수준이라...
    평가지표? 달성률? 이런 거 같은데

    def get_metrics(metrics_config):
    """Get a dictionary of requested metrics instances"""
        return dict((key, globals()[m]())
            for (key, m) in metrics_config.items())

**Comment by ks1101** [comment]
4. trianer.train    //진짜 학습?

train(train_loader, n_epoch = 32, valid_loader)   (trainers/base.py)
    
    #determine starting n_epoch
        기본 0, 하던 작업 있으면 거기서부터 시작
    for(n_epoch = 32)
        train_epoch(train_data_loader)  (trainers/auto_regressive.py)
            self.model.train()  //디버그 모드로 돌려봤는데 뭐가 변하는지 모르겠습니다.
            for(data_loader의 batch 개수)   //2560개입니다. 10240/4
                //data_loader의 batch 데이터는 datasets/climate.py의 __len__, __getitem__에서 가져옵니다.
                //__len__에서는 아마 batch의 길이? 16을 뱉음
                //__getitem__(index)인데 index 값이 기준이 있는지 무작위인지는 모르겠습니다.                     //볼 때마다 다르고 규칙이 안 보여서 무작위로 추정하고 있습니다.
                //__getitem__은 4번 진행합니다.
                //그래서 batch의 크기는 (4, 16, 16, 17, 30) 
                //4 = __getitem__을 4회해서? batch_size?
                //16 = __len__, 학습을 위해 잘라온 데이터의 길이?
                //16, 17, 30 원래 데이터

                batch -> (4, 16, 16, 17, 30)을 batch_input, batch_target으로 분리
                batch_input -> (4,15,16,17,30)  //batch에서 0-14까지의 시간
                batch_target -> (4,15,16,17,30) //batch에서 1-15까지의 시간
                batch_output = model(batch_input)   //여기서 predrnn_pp로 가는데 따로 쓰겠습니다. 언제가 될 지는 모르지만
                //0-14 시간의 데이터로 학습해서 1-15 시간의 데이터를 예측
                loss_func => batch_output, batch_target을 비교함
                이하 backward, optimizer.step등은 더 잘 아실테니까 생략하겠습니다. 찾아보기 귀찮은 건 아닐걸요?

            학습 후
            return loss, metrics_summary
        그 다음에 valid_dataloader로 evaluate() 돌려서 평가하는데
        train_epoch랑  batch_loss.backward(), self.optimizer.step() 말고는 같아서 생략하겠습니다.
    학습 다하면 summaries 반환

**Comment by ks1101** [comment]
5. predrnn_pp
여기는 이해가 안 가는 부분이 많아서 더 볼 필요가 있을 것 같습니다.
batch_output = model(batch_input)   (trainers/auto_regressive.py 33)
    forward (models/preddrnn_pp.py)
        h,c,m,z 초기화하고
        for(batch_input의 시간 = 15)
            시간 단위로 잘라서 clstm 적용 //clstm = causal lstm stack
                clstm 
                filter_size = 3
                num_dims = 2
                channels = [8,16,32,16]
                num_layers = len(channels) = 4
                num_layers가 lstm을 몇 단을 쌓았는지 나타내는 것 같습니다.

                x 크기 -> (4, 16, 17, 30)               //input
                h 크기 -> (4, [8,16,32,16], 17, 30)     //hidden state
                c 크기 -> (4, [8,16,32,16], 17, 30)     //temoral
                m 크기 -> (4, 16, 17, 30)               //spatial
                z 크기 -> (4, 8, 17, 30)                //ghu

                한 번에 한 시간단위만 학습 

                casual lstm, ghu에서는 식은 얼추 맞는 것 같은데
                conv_. 이 어떤 역할을 하는지 모르겠습니다.
                코드에 weight가 안 나와 있어서 weight로 예상을 하고는 있는데 정확히는 모르겠네요.
                conv_.을 거쳐도 데이터 사이즈는 안 바껴서 잘 모르겠습니다

당장 모르겠는 부분
1. h[1], c[1]이 왜 지멋대로인지 //이게 channels 값이긴 한데 어떻게 활용하는 지를 몰라서...
2.conv_. 역할은?
